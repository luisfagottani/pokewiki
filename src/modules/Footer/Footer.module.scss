.footer {
  // Visual
  border-top: 1px solid var(--iron);

  // Box Model
  padding: 25px 24px;
  margin-bottom: 30px;
  max-width: 1200px;
  margin: 0 auto;
  display: flex;
  align-items: center;
  justify-content: space-between;

  &__madeby {
    // Typography
    font-size: 12px;
    line-height: 12px;
    text-align: center;
    color: rgb(204, 200, 200);
  }

  &__translations {
    // Box Model
    display: flex;
    align-items: center;
  }

  &__language {
    // Typography
    font-size: 12px;
    font-weight: bold;
    color: var(--storm-gray);

    // Box Model
    margin-right: 10px;

    &--hide {
      // Box Model
      display: none;
    }

    &:hover {
      // Typography
      text-decoration: underline;
      cursor: pointer;
    }
  }
}
