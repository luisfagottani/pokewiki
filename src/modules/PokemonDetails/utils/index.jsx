const progressBarColor = (progress) => {
  if (progress < 60) {
    return 'blue';
  } else if (progress >= 60 && progress <= 70) {
    return 'yellow';
  } else {
    return 'red';
  }
};

export { progressBarColor };
