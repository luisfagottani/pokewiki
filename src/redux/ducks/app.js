import produce from 'immer';

export const REDUCER_NAME = 'app';

// TYPES
export const Types = {
  SET_GLOBAL_LOADING: `${REDUCER_NAME}/SET_GLOBAL_LOADING`,
};

// INITIAL STATE
const initialState = {
  isLoading: true,
  locale: 'pt-BR',
};

// REDUCER
export const reducer = produce((draft = initialState, action) => {
  switch (action.type) {
    case Types.SET_GLOBAL_LOADING:
      draft.isLoading = action.payload;
      return draft;

    case Types.SET_USER_PROFILE:
      draft.profile = action.profile;
      return draft;

    default:
      return draft;
  }
});

// ACTION FUNCTIONS

export function defineGlobalLoading(status) {
  return {
    type: Types.SET_GLOBAL_LOADING,
    payload: status,
  };
}

// SELECTOR FUNCTIONS
export function getGlobalLoading(state) {
  return state[REDUCER_NAME].isLoading;
}

export function getLocale(state) {
  return state[REDUCER_NAME].locale;
}
