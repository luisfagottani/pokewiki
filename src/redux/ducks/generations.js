import produce from 'immer';

export const REDUCER_NAME = 'family';

// TYPES
export const Types = {
  FETCH_GENERATIONS: `${REDUCER_NAME}/FETCH_POKEMONS`,
  FETCH_GENERATIONS_BY_ID: `${REDUCER_NAME}/FETCH_GENERATIONS_BY_ID`,
  PERSIST_GENERATIONS: `${REDUCER_NAME}/PERSIST_GENERATIONS`,
};

// INITIAL STATE
const initialState = {
  generations: [],
};

// REDUCER
export const reducer = produce((draft = initialState, action) => {
  switch (action.type) {
    case Types.PERSIST_GENERATIONS:
      draft.generations = { ...action.payload };
      return draft;

    default:
      return draft;
  }
});

// ACTION FUNCTIONS
export function fetchGenerations() {
  return {
    type: Types.FETCH_GENERATIONS,
  };
}

export function fetchGenerationsById(id) {
  return {
    type: Types.FETCH_GENERATIONS_BY_ID,
    payload: id,
  };
}

export function persistGenerations(pokemons) {
  return {
    type: Types.PERSIST_GENERATIONS,
    payload: pokemons,
  };
}

// SELECTORS
export function getGenerations(state) {
  return state[REDUCER_NAME]?.generations;
}
