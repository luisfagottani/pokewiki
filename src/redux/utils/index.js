const reduceSpecies = (results) =>
  results.reduce((current, next) => {
    const name = next?.name;
    if (!current[name]) {
      current[name] = { name };
    }
    return current;
  }, {});

export { reduceSpecies };
