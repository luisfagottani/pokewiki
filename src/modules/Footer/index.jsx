import classNames from 'classnames';
import { useIntl } from 'react-intl';
import { connect } from 'react-redux';

import { defineLocale, getLocale } from 'redux/ducks/app';

import style from './Footer.module.scss';

const Footer = ({ currentLocale, handleDefineLocale }) => {
  const defineLocaleOnClick = (locale) => {
    handleDefineLocale(locale);
  };

  const intl = useIntl();
  return (
    <footer className={style.footer}>
      <ul className={style.footer__translations}>
        <li
          onClick={() => defineLocaleOnClick('pt-BR')}
          className={classNames(style.footer__language, {
            [style['footer__language--hide']]: currentLocale === 'pt-BR',
          })}
        >
          PT-BR
        </li>
        <li
          onClick={() => defineLocaleOnClick('EN')}
          className={classNames(style.footer__language, { [style['footer__language--hide']]: currentLocale === 'EN' })}
        >
          EN
        </li>
      </ul>
      <p className={style.footer__madeby}>{intl.formatMessage({ id: 'footer.description' })}</p>
    </footer>
  );
};

const mapStateToProps = (state) => ({
  currentLocale: getLocale(state),
});

const mapDispatchToProps = {
  handleDefineLocale: defineLocale,
};

export default connect(mapStateToProps, mapDispatchToProps)(Footer);
