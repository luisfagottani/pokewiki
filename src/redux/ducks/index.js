import { combineReducers } from 'redux';

import { reducer as AppReducer, REDUCER_NAME as APP_REDUCER_NAME } from './app';
import { reducer as PokemonsReducer, REDUCER_NAME as APP_POKEMONS_REDUCER } from './pokemons';
import { reducer as GenerationsReducer, REDUCER_NAME as APP_GENERATIONS_REDUCER } from './generations';

export default combineReducers({
  [APP_REDUCER_NAME]: AppReducer,
  [APP_POKEMONS_REDUCER]: PokemonsReducer,
  [APP_GENERATIONS_REDUCER]: GenerationsReducer,
});
