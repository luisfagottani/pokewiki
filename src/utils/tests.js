import React from 'react';
import { render as rtlRender } from '@testing-library/react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { IntlProvider } from 'react-intl';
import { createStore, combineReducers } from 'redux';

import { Messages } from 'i18n';
import middleware, { sagaMiddleware } from 'redux/middleware';

export const componentWithRouter = (initialRoute) => (children) => {
  if (typeof children === 'function') {
    return (
      <MemoryRouter initialIndex={0} initialEntries={initialRoute && [initialRoute]}>
        {children()}
      </MemoryRouter>
    );
  }
  return (
    <MemoryRouter initialEntries={initialRoute && [initialRoute]} initialIndex={0}>
      {children}
    </MemoryRouter>
  );
};

export const createMockStore = ({ reducers, initialState }) => {
  return createStore(combineReducers(reducers), initialState, middleware);
};

export const mockSaga = (fnGenerators) => sagaMiddleware.run(fnGenerators);

function render(ui, store, { ...renderOptions } = {}) {
  const locale = store.getState().app.locale ?? 'EN';
  function Wrapper({ children }) {
    return (
      <Provider store={store}>
        <IntlProvider messages={Messages[locale]} locale={locale}>
          {children}
        </IntlProvider>
      </Provider>
    );
  }

  return { store, ...rtlRender(ui, { wrapper: Wrapper, ...renderOptions }) };
}

// re-export everything
export * from '@testing-library/react';

// override render method
export { render };
