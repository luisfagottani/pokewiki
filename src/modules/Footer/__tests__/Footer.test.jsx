import React from 'react';
import { createMockStore, render, fireEvent } from 'utils/tests';
import { reducer as AppReducer, REDUCER_NAME as APP_REDUCER_NAME } from 'redux/ducks/app';
import { act } from 'react-dom/test-utils';
import Footer from '..';
import store from 'redux';

const mockStore = createMockStore({
  reducers: { [APP_REDUCER_NAME]: AppReducer },
  initialState: {
    [APP_REDUCER_NAME]: {
      isLoading: true,
      locale: 'EN',
    },
  },
});

describe('<Footer />', () => {
  it('Should render without errors', () => {
    const { container } = render(<Footer />, mockStore);
    expect(container).toMatchSnapshot();
  });

  describe('When the english is the current locale', () => {
    it('should render the pt-br option', () => {
      const { getByText } = render(<Footer />, mockStore);

      getByText('PT-BR');
    });
  });

  describe('When the pt-br is the current locale', () => {
    it('should render the en option', () => {
      const { getByText } = render(<Footer />, mockStore);

      getByText('EN');
    });
  });

  describe('When click to change the locale', () => {
    it('should render the correct option', async () => {
      const { getByText, store } = render(<Footer />, mockStore);
      getByText('PokeAPI by Luis Agottani with CRA');

      const ptBrButton = getByText('PT-BR');

      await act(async () => {
        fireEvent.click(ptBrButton, { button: 1 });
      });

      expect(store.getState().app.locale).toBe('pt-BR');
    });
  });
});
