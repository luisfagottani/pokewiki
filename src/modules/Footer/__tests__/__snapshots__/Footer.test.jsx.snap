// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Footer /> Should render without errors 1`] = `
<div>
  <footer
    class="footer"
  >
    <ul
      class="footer__translations"
    >
      <li
        class="footer__language"
      >
        PT-BR
      </li>
      <li
        class="footer__language footer__language--hide"
      >
        EN
      </li>
    </ul>
    <p
      class="footer__madeby"
    >
      PokeAPI by Luis Agottani with CRA
    </p>
  </footer>
</div>
`;
